src/ripple/app/paths/impl/BookStep.cpp

```diff
@@ -483,35 +481,82 @@ BookStep<TIn, TOut>::fwdImp (
             std::uint32_t transferRateIn,
             std::uint32_t transferRateOut) mutable -> bool
     {
+        assert(cache_);
+
         if (remainingIn <= beast::zero)
             return false;
 
+        bool processMore = true;
+        auto ofrAdjAmt = ofrAmt;
+        auto stpAdjAmt = stpAmt;
+        auto ownerGivesAdj = ownerGives;
+
+        typename boost::container::flat_multiset<TOut>::const_iterator lastOut;
         if (stpAmt.in <= remainingIn)
         {
             savedIns.insert(stpAmt.in);
-            savedOuts.insert(stpAmt.out);
+            lastOut = savedOuts.insert(stpAmt.out);
             result = TAmounts<TIn, TOut>(sum (savedIns), sum(savedOuts));
-            remainingIn = in - result.in;
-            this->consumeOffer (sb, offer, ofrAmt, stpAmt, ownerGives);
-            // return true b/c even if the payment is satisfied,
-            // we need to consume the offer
-            return true;
+            // consume the offer even if stepAmt.in == remainingIn
+            processMore = true;
         }
         else
         {
-            auto ofrAdjAmt = ofrAmt;
-            auto stpAdjAmt = stpAmt;
-            auto ownerGivesAdj = ownerGives;
             limitStepIn (offer.quality (), ofrAdjAmt, stpAdjAmt, ownerGivesAdj,
                 transferRateIn, transferRateOut, remainingIn);
             savedIns.insert (remainingIn);
-            savedOuts.insert (stpAdjAmt.out);
-            remainingIn = beast::zero;
+            lastOut = savedOuts.insert (stpAdjAmt.out);
             result.out = sum (savedOuts);
             result.in = in;
-            this->consumeOffer (sb, offer, ofrAdjAmt, stpAdjAmt, ownerGivesAdj);
-            return false;
+
+            processMore = false;
         }
+
+        if (result.out > cache_->out && result.in <= cache_->in)
+        {
+            // The step produced more output in the forward pass than the
+            // reverse pass while consuming the same input (or less). If we
+            // compute the input required to produce the cached output
+            // (produced in the reverse step) and the input is equal to
+            // the input consumed in the forward step, then consume the
+            // input provided in the forward step and produce the output
+            // requested from the reverse step.
+            auto const lastOutAmt = *lastOut;
```

Comment:
Woah.  Great paragraph, but it reads in a way that was slightly startling to me.  I eventually decided that it's not really very startling.  I think this is what you're doing:
 1. If the reverse pass and the forward pass consume the same input,
 2. And the forward pass produces more output than the reverse pass,
 3. Then use the reverse pass's (smaller) output value.

Did I get that right?